### ISSUE TITLE
Some recommended alternatives to useEffect are less declarative and concise

### ISSUE THREAD

**Issue by jedwards1211** [reactjs/react.dev#7012]
### Summary

You Might Not Need an Effect has this example:
```js
  // 🔴 Avoid: Adjusting state on prop change in an Effect
  useEffect(() => {
    setSelection(null);
  }, [items]);
```
I understand the technical reasons for discouraging `useEffect`, but it seems like the authors haven't considered this from a readability perspective.

The docs suggest this instead, which is way less concise, declarative, and IMO much less readable:
```js
  // Better: Adjust the state while rendering
  const [prevItems, setPrevItems] = useState(items);
  if (items !== prevItems) {
    setPrevItems(items);
    setSelection(null);
  }
```

### Page

https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes

### Details

To have the better readability, I would use this custom hook in userland.  Since it's so easy for people to abuse `useEffect`, I would argue React should provide something like the following as a builtin.  But if not, the docs should at least recommend it:

```js
useImmediateEffect(() => {
  setSelection(null);
}, [items]);
```
```js
function useImmediateEffect(effect, deps) {
  const [finishLastEffect, setFinishLastEffect] = useState(() => {});
  const [prevDeps, setPrevDeps] = useState(deps);
  if (!shallowEqual(deps, prevDeps)) {
    setPrevDeps(deps);
    finishLastEffect();
    setFinishLastEffect(effect());
  }
}

```

**Comment by Luccasoli** [comment]
I agree! And it helps to influence technically within the technical teams we work with!

I recently saw this article creating a custom hook similar to yours: https://www.jacobparis.com/content/use-reset-callback#extracting-the-logic-into-a-custom-hook

````jsx
function useResetCallback(initialValue, resetFn: () => any) {
  const [prevValue, setPrevValue] = useState(initialValue)
  if (prevValue !== initialValue) {
    resetFn()
    setPrevValue(initialValue)
  }
}
````

**Comment by karlhorky** [comment]
Also relevant here are two new React Compiler ESLint rules:

- [`react-hooks/set-state-in-effect`](https://react.dev/reference/eslint-plugin-react-hooks/lints/set-state-in-effect)
- [`react-hooks/set-state-in-render`](https://react.dev/reference/eslint-plugin-react-hooks/lints/set-state-in-render)

Enabling and following the problems reported by these two lint rules makes it much more challenging to achieve deriving state, as synchronous `setState()` usage in `useEffect()` or render will now be reported as a problem.

**Comment by jedwards1211** [comment]
@karlhorky okay I tried out `set-state-in-render` and it seems to ignore `setState` calls inside a conditional like
```tsx
  if (items !== prevItems) {
    setPrevItems(items);
  }
```
so thank goodness for that.  That does mean it wouldn't catch infinite loops like the following though, so we'll see if it stays that way
```tsx
  if (items !== prevItems) {
    setPrevItems([...items]);
  }
```

**Comment by gaearon** [comment]
The linter plugin works correctly but its docs were misleading. I edited in https://github.com/reactjs/react.dev/pull/8033/files.

Re: the original point, yes, this pattern is annoying. Ideally there would be a first-class API for this that is less awkward to use. The team is aware of the pain point.

I would not recommend the custom hook you're suggesting. It hides the pattern away and makes excessive render cycles invisible. It's also not an effect and it is misleading in the sense that people *will* start putting arbitrary side effects there. Whereas the only thing you're allowed to do during the render cycle is to call `setState`. So the awkward raw pattern is better. If there's ever a first-class API for it, it would be easy to search for and codemod.

**Comment by jedwards1211** [comment]
> makes excessive render cycles invisible

The tension between simple, declarative code and good rendering performance is pretty fundamental to all of React.  Is there really a coherent philosophy about what should and shouldn't be hidden?  I think declarative patterns are still worthwhile in many cases

**Comment by gaearon** [comment]
I agree that this *does* need a declarative API. Just not what's suggested in this thread (because it's not an "effect", doesn't behave like one, and it would be bad if people started putting things into their bodies that they usually put into "effects").

A declarative API for this could look like modifying `useState` to have an extra argument that lets you reset it or something like that. I do agree it's worthwhile, just not this one.

**Comment by jedwards1211** [comment]
Would it be limited to resetting to the initial value though?  I would hope for something more flexible

**Comment by jedwards1211** [comment]
For example I have the following in one of my components, and I wish there were a more concise best practice way of declaring what should happen here:

```tsx
export function DebugPlots({ config, ranges: inputRanges }) {
  const [ranges, setRanges] = useState(inputRanges || {})

  const autoTimeRange = useMemo(() => {
    if (inputRanges?.time) return undefined
    ... // (computes a luxon time Interval from config.plots if they have data available)
  }, [config.plots]);

  // apply the autoTimeRange to state whenever it changes, as long as it's fully defined

  const [lastAutoTimeRange, setLastAutoTimeRange] = useState(autoTimeRange);
  if (
    autoTimeRange &&
    (lastAutoTimeRange == null || !autoTimeRange.equals(lastAutoTimeRange))
  ) {
    setLastAutoTimeRange(autoTimeRange);
    const { start, end } = autoTimeRange;
    if (start && end) {
      setRanges((ranges) => ({
        ...ranges, time: { left: +start, right: +end } },
      }));
    }
  }

  ...

  // (event handlers may call setRanges for other axes besides time)
}
```

**Comment by gaearon** [comment]
When you have multiple pieces of state resetting each other, you usually want to move that into a reducer. Does that help?

**Comment by jedwards1211** [comment]
Not really...the resetting only happens in one direction, not both.  The second `useState` call is just helping determine if the auto time range has changed from one render to another.  The docs recommend using `useState` to check if a value is different from the previous render, not `useReducer`.

This is a plot component where the user can drag the time and value range around with the mouse.  For convenience, the developer can either pass in the initial time range they want, or omit that prop and let the component choose an auto time range from the data series.  But after that, the user can manipulate the time range.  If a new data series is passed into the plot though, and no time range is passed in, it should reset to the auto time range for the new data.
Using a reducer for this would be more cumbersome to me

**Comment by karlhorky** [comment]
I ran into a [`react-hooks/set-state-in-effect`](https://react.dev/reference/eslint-plugin-react-hooks/lints/set-state-in-effect) error with [`useActionState()`](https://react.dev/reference/react/useActionState) (the line after `eslint-disable-next-line`):

```tsx
function UpdateUserForm(props: Props) {
  const [state, formAction, isPending] = useActionState(updateUserAction, {
    type: 'initial',
  });
  const [isEditing, setIsEditing] = useState(false);

  // Reset editing after user updated successfully
  useEffect(() => {
    // eslint-disable-next-line react-hooks/set-state-in-effect -- TODO: Consider how to do this without setState in useEffect, maybe with setting state during render https://react.dev/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state:~:text=adjust%20the%20state%20directly%20during%20rendering
    if (state.type === 'success') setIsEditing(false);
  }, [state]);

  return (
    <form action={formAction}>
      <EditableField
        isEditing={isEditing}
        name="firstName"
        defaultValue={props.user.firstName}
      />
      <EditableField
        isEditing={isEditing}
        name="lastName"
        defaultValue={props.user.lastName}
      />

      {!isEditing ? (
        <button
          onClick={(event) => {
            event.preventDefault();
            setIsEditing(true);
          }}
        >
          Edit
        </button>
      ) : (
        <button disabled={isPending}>Save</button>
      )}

      <ServerActionErrors state={state} />
    </form>
  );
}
```

Based on [Storing information from previous renders](https://react.dev/reference/react/useState#storing-information-from-previous-renders) and [Adjusting some state when a prop changes](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes), the previous state could be saved and state set during render instead:

```diff
+  const [prevState, setPrevState] = useState(state);

   // Reset editing after user updated successfully
-  useEffect(() => {
-    // eslint-disable-next-line react-hooks/set-state-in-effect -- TODO: Consider how to do this without setState in useEffect, maybe with setting state during render https://react.dev/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state:~:text=adjust%20the%20state%20directly%20during%20rendering
-    if (state.type === 'success') setIsEditing(false);
-  }, [state]);
+  if (previousState !== state) {
+    setPreviousState(state);
+    if (state.type === 'success') setIsEditing(false);
+  }
```

This works, so I'm assuming this is the recommended pattern here: trading the `useEffect()` for an additional state variable.

### Alternatives considered

1. Use [the `key` prop](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes:~:text=each%20user%E2%80%99s%20profile%20is%20conceptually%20a%20different%20profile%20by%20giving%20it%20an%20explicit%20key) to re-render everything in the component (downsides: re-renders everything, possibly losing form field focus)
2. Use an async function instead of passing `updateUserAction` directly to `useActionState()` (downside: gives up [progressive enhancement](https://nextjs.org/docs/app/getting-started/updating-data#server-components:~:text=Good%20to%20know%3A%20Server%20Components%20support%20progressive%20enhancement%20by%20default%2C%20meaning%20forms%20that%20call%20Server%20Actions%20will%20be%20submitted%20even%20if%20JavaScript%20hasn%27t%20loaded%20yet%20or%20is%20disabled.))
3. Moving the `isEditing` state variable into the action state (downside: it doesn't belong in action state)
4. [Using `startTransition()`, `requestAnimationFrame()` or `setTimeout()`](https://github.com/reactjs/react.dev/pull/8029#discussion_r2394023573) around the `setIsEditing()` in the `useEffect()` (downside: seems like tricking the lint rule)

**Comment by jedwards1211** [comment]
Isn't all you need in this case
```
if (state.type === 'success' && isEditing) setIsEditing(false);
```

Honestly though I don't get why a `setState` should cause a rerender if the new state value is equal to the current

**Comment by jedwards1211** [comment]
Honestly I would be tempted to make a hook like

```
if (useIsChanged(state.type)) {
  ...
}
```

were it not for the fact that it would be too easy to make it conditionally run:

```
if (state.type === 'success' && useIsChanged(state.type)) { // oops
```

**Comment by jedwards1211** [comment]
@gaearon 
> because it's not an "effect", doesn't behave like one

The need to "do X if Y changed" comes up all the time for many of us using React.

*React's design causes us to think in these terms* since components are render functions that get called anytime time new props have been passed or local state is changed.

We have, if I understand correctly
- "do X in a later event queue event if Y changed" (`useEffect`)
- "do X synchronously after render if Y changed" (`useLayoutEffect`)

What we don't have is the conceptually similar
- "do X right now if Y changed"

Which is personally the only thing I'd care for React to add lately.

You may not think it qualifies as an "effect" but you can see the conceptual similarity of "do X at &lt;time&gt; when Y changed" right?  What else can we call this?  A "trigger" maybe?

I just want a way to *express my thoughts fluently within the way React's design forces me to think.*

What we really need is an API for "if Y changed" that's not cumbersome or error-prone.

**Comment by aweebit** [comment]
> I agree that this _does_ need a declarative API. Just not what's suggested in this thread (because it's not an "effect", doesn't behave like one, and it would be bad if people started putting things into their bodies that they usually put into "effects").
> 
> A declarative API for this could look like modifying `useState` to have an extra argument that lets you reset it or something like that. I do agree it's worthwhile, just not this one.

@gaearon There is an open issue in the main repo that suggests exactly that. Please have a look at it if you haven't seen it yet:

- https://github.com/facebook/react/issues/33041

**Comment by aweebit** [comment]
> For example I have the following in one of my components, and I wish there were a more concise best practice way of declaring what should happen here:
> 
> ```ts
> export function DebugPlots({ config, ranges: inputRanges }) {
>   const [ranges, setRanges] = useState(inputRanges || {})
> 
>   const autoTimeRange = useMemo(() => {
>     if (inputRanges?.time) return undefined
>     ... // (computes a luxon time Interval from config.plots if they have data available)
>   }, [config.plots]);
> 
>   // apply the autoTimeRange to state whenever it changes, as long as it's fully defined
> 
>   const [lastAutoTimeRange, setLastAutoTimeRange] = useState(autoTimeRange);
>   if (
>     autoTimeRange &&
>     (lastAutoTimeRange == null || !autoTimeRange.equals(lastAutoTimeRange))
>   ) {
>     setLastAutoTimeRange(autoTimeRange);
>     const { start, end } = autoTimeRange;
>     if (start && end) {
>       setRanges((ranges) => ({
>         ...ranges, time: { left: +start, right: +end } },
>       }));
>     }
>   }
> 
>   ...
> 
>   // (event handlers may call setRanges for other axes besides time)
> }
> ```

@jedwards1211 Something like this would be a solution to your problem if React team was to implement my suggestion from https://github.com/facebook/react/issues/33041:

```ts
const [ranges, setRanges] = useState((prev) => {
  if (inputRanges) return inputRanges;
  const { start, end } = computeAutoTimeRange(config.plots);
  if (start && end) {
    const {
      time: { left: prevLeft, right: prevRight },
    } = { time: {}, ...prev };
    return +start === prevLeft && +end === prevRight
      ? prev
      : { ...prev, time: { left: +start, right: +end } };
  }
  return prev ?? {};
}, [inputRanges, config.plots]);
```

In my [@aweebit/react-essentials](https://www.npmjs.com/package/@aweebit/react-essentials) library, there is a `useStateWithDeps` hook available that encapsulates this functionality, so you can try this feature out today even without React supporting it out of the box.

**Comment by aweebit** [comment]
> I ran into a [`react-hooks/set-state-in-effect`](https://react.dev/reference/eslint-plugin-react-hooks/lints/set-state-in-effect) error with [`useActionState()`](https://react.dev/reference/react/useActionState) (the line after `eslint-disable-next-line`):
>
> ```tsx
> function UpdateUserForm(props: Props) {
>   const [state, formAction, isPending] = useActionState(updateUserAction, {
>     type: 'initial',
>   });
>   const [isEditing, setIsEditing] = useState(false);
> 
>   // Reset editing after user updated successfully
>   useEffect(() => {
>     // eslint-disable-next-line react-hooks/set-state-in-effect -- TODO: Consider how to do this without setState in useEffect, maybe with setting state during render https://react.dev/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state:~:text=adjust%20the%20state%20directly%20during%20rendering
>     if (state.type === 'success') setIsEditing(false);
>   }, [state]);
> 
>   return (
>     <form action={formAction}>
>       <EditableField
>         isEditing={isEditing}
>         name="firstName"
>         defaultValue={props.user.firstName}
>       />
>       <EditableField
>         isEditing={isEditing}
>         name="lastName"
>         defaultValue={props.user.lastName}
>       />
> 
>       {!isEditing ? (
>         <button
>           onClick={(event) => {
>             event.preventDefault();
>             setIsEditing(true);
>           }}
>         >
>           Edit
>         </button>
>       ) : (
>         <button disabled={isPending}>Save</button>
>       )}
> 
>       <ServerActionErrors state={state} />
>     </form>
>   );
> }
> ```
> 
> Based on [Storing information from previous renders](https://react.dev/reference/react/useState#storing-information-from-previous-renders) and [Adjusting some state when a prop changes](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes), the previous state could be saved and state set during render instead:
>
> ```diff
> +  const [prevState, setPrevState] = useState(state);
> 
>    // Reset editing after user updated successfully
> -  useEffect(() => {
> -    // eslint-disable-next-line react-hooks/set-state-in-effect -- TODO: Consider how to do this without setState in useEffect, maybe with setting state during render https://react.dev/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state:~:text=adjust%20the%20state%20directly%20during%20rendering
> -    if (state.type === 'success') setIsEditing(false);
> -  }, [state]);
> +  if (previousState !== state) {
> +    setPreviousState(state);
> +    if (state.type === 'success') setIsEditing(false);
> +  }
> ```
> 
> This works, so I'm assuming this is the recommended pattern here: trading the `useEffect()` for an additional state variable.
> ### Alternatives considered
> 
> 1. Use [the `key` prop](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes:~:text=each%20user%E2%80%99s%20profile%20is%20conceptually%20a%20different%20profile%20by%20giving%20it%20an%20explicit%20key) to re-render everything in the component (downsides: re-renders everything, possibly losing form field focus)
> 2. Use an async function instead of passing `updateUserAction` directly to `useActionState()` (downside: gives up [progressive enhancement](https://nextjs.org/docs/app/getting-started/updating-data#server-components:~:text=Good%20to%20know%3A%20Server%20Components%20support%20progressive%20enhancement%20by%20default%2C%20meaning%20forms%20that%20call%20Server%20Actions%20will%20be%20submitted%20even%20if%20JavaScript%20hasn%27t%20loaded%20yet%20or%20is%20disabled.))
> 3. Moving the `isEditing` state variable into the action state (downside: it doesn't belong in action state)
> 4. [Using `startTransition()`, `requestAnimationFrame()` or `setTimeout()`](https://github.com/reactjs/react.dev/pull/8029#discussion_r2394023573) around the `setIsEditing()` in the `useEffect()` (downside: seems like tricking the lint rule)

Also for this there would a very simple and elegant solution if React was to offer the functionality I proposed in https://github.com/facebook/react/issues/33041 out of the box:

```diff
-   const [isEditing, setIsEditing] = useState(false);
-
-   // Reset editing after user updated successfully
-   useEffect(() => {
-     // eslint-disable-next-line react-hooks/set-state-in-effect -- TODO: Consider how to do this without setState in useEffect, maybe with setting state during render https://react.dev/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state:~:text=adjust%20the%20state%20directly%20during%20rendering
-     if (state.type === 'success') setIsEditing(false);
-   }, [state]);
+   const [isEditing, setIsEditing] = useState<boolean>(
+     (prev) => state.type === 'success' ? false : prev ?? false,
+     [state.type],
+   );
```

**Comment by jedwards1211** [comment]
Doesn't Dan's complaint that

> It hides the pattern away and makes excessive render cycles invisible

Apply just as much to this `useStateWithDeps` pattern as to my `useImmediateEffect`?  If there are excessive render cycles because of deps changing too often, the debug experience is going to be pretty much the same for both of these.

And although `useStateWithDeps` would be sufficient for many of my use cases, it's still not as flexible as a general purpose "do X immediately if Y changed" function.

People didn't misuse `useEffect` because they're stupid; they misused it because "do X if Y changed" is a crystal clear concept.  The only non-obvious part is the timing of when X gets done.

**Comment by aweebit** [comment]
> Doesn't Dan's complaint that
> 
> > It hides the pattern away and makes excessive render cycles invisible
> 
> Apply just as much to this `useStateWithDeps` pattern as to my `useImmediateEffect`? If there are excessive render cycles because of deps changing too often, the debug experience is going to be pretty much the same for both of these.

It does if we're talking about my user-land implementation of `useStateWithDeps`, but it wouldn't if the functionality was offered by React out of the box, which is part of the reason why I'm pushing for this change so much.

You see, the only way to correctly implement `useStateWithDeps` in user-land without breaking the rules of React is to rely on that same `prevState` pattern that you rely on in your implementation of `useImmediateEffect`, but that pattern is exactly the thing that causes those excessive render cycles Dan was talking about to happen.

But if React itself was to offer the functionality, there would be no excessive rendering whatsoever because the resetting would take place internally during `useState` execution before it even returns, so your state variable's value would be correct already during the first render caused by a change in dependencies, and because of that no second render would be necessary.

> And although `useStateWithDeps` would be sufficient for many of my use cases, it's still not as flexible as a general purpose "do X immediately if Y changed" function.

But is it actually common that neither `useStateWithDeps` nor [`useReducerWithDeps`](https://github.com/aweebit/react-essentials?tab=readme-ov-file#usereducerwithdeps) (its reducer pattern counterpart) can be used? I mean, is it common enough for React to include something like `useImmediateEffect` that could cause a lot of confusion and misuse in its public API? I would like to see your examples.

**Comment by jedwards1211** [comment]
> It does if we're talking about my user-land implementation of useStateWithDeps, but it wouldn't if the functionality was offered by React out of the box

Yes of course, but by this same token if React built in a "do X immediately when Y changed" API along the lines of what I'm proposing, it could also provide a better debug experience.  Neither API has an advantage in this regard.

> But is it actually common that neither useStateWithDeps nor [useReducerWithDeps](https://github.com/aweebit/react-essentials?tab=readme-ov-file#usereducerwithdeps) (its reducer pattern counterpart) can be used?

It's probably not that common, no.  But flexibility is power -- a single "do X immediately when Y changed" API would satisfy the needs for both `useState` and `useReducer` without having to create new separate `useStateWithDeps` and `useReducerWithDeps` APIs.

Again, if we have

- "do X in a later event queue event if Y changed" (useEffect)
- "do X synchronously after render if Y changed" (useLayoutEffect)

What's the problem with

- "do X immediately if Y changed"?

**Comment by jedwards1211** [comment]
> But if React itself was to offer the functionality, there would be no excessive rendering whatsoever because the resetting would take place internally during useState execution before it even returns

I was thinking, this isn't going to make it any more obvious if one of the deps isn't correctly memoized and changes every render.  But in that case what I'm proposing would render twice before commit, whereas `useStateWithDeps` would only render once before commit -- I guess that's the excessive rendering you're focusing on?

**Comment by aweebit** [comment]
> It's probably not that common, no. But flexibility is power -- a single "do X immediately when Y changed" API would satisfy the needs for both `useState` and `useReducer` without having to create new separate `useStateWithDeps` and `useReducerWithDeps` APIs.

No separate new APIs should be created when React adds this – just the existing ones each extended by a dependency array parameter.

> Yes of course, but by this same token if React built in a "do X immediately when Y changed" API along the lines of what I'm proposing, it could also provide a better debug experience.

I'm not sure what exactly you mean here. How would `useImmediateEffect` be better in terms of debugging experience? And if it actually is, is it really worth it if the cost is having to do an unnecessary second render whose results are immediately thrown away whenever a dependency changes?

> Again, if we have
> * "do X in a later event queue event if Y changed" (useEffect)
> * "do X synchronously after render if Y changed" (useLayoutEffect)
>
> What's the problem with
> * "do X immediately if Y changed"?

The problem is that people will misuse it as a simple escape hatch for whatever they're doing without realizing its downsides. They already do that with `useEffect` despite React team trying hard to discourage such anti-patterns.

`useEffect` cannot be left out of the library though, but `useImmediateEffect` can and probably should remain left out, especially since, as you said yourself, most of its use cases would already be covered by `useState` / `useReducer` dependency arrays.

> I was thinking, this isn't going to make it any more obvious if one of the deps isn't correctly memoized and changes every render.

The [`react-hooks/exhaustive-deps`](https://react.dev/reference/eslint-plugin-react-hooks/lints/exhaustive-deps) rule already does a good job at preventing such problems for `useEffect`, `useMemo`, etc., and I don't see any reason why it couldn't be adjusted to also report them for `useState` / `useReducer` dependencies in just the same way.

> But in that case what I'm proposing would render twice before commit, whereas `useStateWithDeps` would only render once before commit -- I guess that's the excessive rendering you're focusing on?

Yeah, with excessive or unnecessary renders I mean those whose results are thrown away immediately because a state setter was called synchronously from within the component's body, as described in [useState > Storing information from previous renders](https://react.dev/reference/react/useState#storing-information-from-previous-renders):

> _The rest of your component function will still execute (and the result will be thrown away)._

That _always_ happens in the `prevState` approach (the only somewhat decent one we currently have) whenever a dependency changes (even if the state that depends on it doesn't), but no such unnecessary renders would have to take place if React was to support `useState` / `useReducer` dependency arrays out of the box.

**Comment by aweebit** [comment]
Actually, this `useImmediateEffect` approach was the first one I came up with when I was looking for an elegant solution to the derived state problem for the first time.

However, after I discovered the idea of `useState` being extended by a dependency array parameter, I realized it was much better and never wanted to come back to `useImmediateEffect` ever again.

I think the lack of flexibility is actually this solution's advantage because it means that it's much harder to abuse it.

And in general, it just matches React's declarative nature much better.

React offers you effects as a way to synchronize your components with external systems, and not as a way to facilitate state management.

**Comment by jedwards1211** [comment]
> Actually, this useImmediateEffect approach was the first one I came up with when I was looking for an elegant solution to the derived state problem for the first time.

Like I've been saying, "set X when Y changed" is the most natural train of thought when it comes to dependent state.

I don't think there's any avoiding that, but I guess `useState` with deps won't be that hard to get used to since it's like a mashup of existing `useState` and `useMemo`.  I can see the benefit of avoiding a render that gets thrown away, that's selling me on this API more than the ergonomics.

I do still wish there were a more straightforward solution to one of my examples above where I need to do

```ts
if (!isDeepEqual(foo, previousFoo)) {
  setState(deriveState(foo))
}
```

With `useState` with deps, it would need to be something like

```ts
const [{derived}, setDerivedAndFoo] = useState(
  (prev) => isDeepEqual(prev.foo, foo) ? prev : { foo, derived: deriveState(foo) },
  [foo]
)
const setDerived = useCallback(
  (newDerived) => setDerivedAndFoo(state => ({
    ...state,
    derived: typeof newDerived === 'function'
      ? newDerived(state.derived)
      : newDerived
  })),
  []
)
```

**Comment by aweebit** [comment]
> Like I've been saying, "set X when Y changed" is the most natural train of thought when it comes to dependent state.

I think it was only natural to me because React had taught me that the only right way to reset / derive state was by imperatively invoking state setters, and so that was why I was looking for a more ergonomic way to do exactly that. But now that I've seen a different way, I think that this pattern is actually the one that should feel unnatural considering how it goes against React's declarative nature.

> I do still wish there were a more straightforward solution to one of my examples above where I need to do
> 
> ```ts
> if (!isDeepEqual(foo, previousFoo)) {
>   setState(deriveState(foo))
> }
> ```
> 
> With `useState` with deps, it would need to be something like
> 
> ```ts
> const [{derived}, setDerivedAndFoo] = useState(
>   (prev) => isDeepEqual(prev.foo, foo) ? prev : { foo, derived: deriveState(foo) },
>   [foo]
> )
> const setDerived = useCallback(
>   (newDerived) => setDerivedAndFoo(state => ({
>     ...state,
>     derived: typeof newDerived === 'function'
>       ? newDerived(state.derived)
>       : newDerived
>   })),
>   []
> )
> ```

Is the idea here that the state should be reset whenever `foo` changes? Then why not simply do this?

```ts
const [derived, setDerived] = useState(() => deriveState(foo), [foo]);
```

Why do you need to store `foo` in that same state? Is it because you want to compare using `isDeepEqual` rather than `Object.is` which React uses internally to compare your dependencies? In that case, I think you're making a mistake somewhere else: `foo`'s value shouldn't be updated at all if the new value is deeply equal to the previous one! What would be the point of such an update? It would just result in another unnecessary render.

To prevent such pointless updates, you could replace code like this

```ts
setFoo({ a, b, c });
```

with this:

```ts
setFoo((prev) => {
  const next = { a, b, c };
  return isDeepEqual(prev, next) ? prev : next;
});
```

**Comment by jedwards1211** [comment]
> the only right way to reset / derive state was by imperatively invoking state setters

Sure it's not the only way, but if you have state, that you can set (imperatively), it's pretty natural to think of state changes as an imperative process.

With a hybrid imperative/declarative API like `useState` with deps, people may have confusion at times thinking "how did this state change, if nothing ever called `setState`?" forgetting that it can also change via the declarative path.

> Is the idea here that the state should be reset whenever foo changes?

No, for example `foo` could be an object-valued property that you don't want the caller to have to go to the trouble of memoizing, but you want to reset the state when they pass a value that's not deep equal to the last.

**Comment by aweebit** [comment]
> With a hybrid imperative/declarative API like `useState` with deps, people may have confusion at times thinking "how did this state change, if nothing ever called `setState`?" forgetting that it can also change via the declarative path.

I think at the beginning it will be hard to forget considering how it will be a new feature that will feel special every time you use it.

And when it's not special anymore because you've got so used to it, you also won't forget it exactly because of how used to it you've got :) I think you will then always have this possibility of state being reset because of a change in dependencies at the back of your mind.

> No, for example `foo` could be an object-valued property that you don't want the caller to have to go to the trouble of memoizing, but you want to reset the state when they pass a value that's not deep equal to the last.

Ah, I see. I actually had a case like that with the `options` parameter of the [`useEventListener`](https://github.com/aweebit/react-essentials/tree/v0.10.6?tab=readme-ov-file#useeventlistener) hook from my library. Here is how I solved it:

```ts
const {
  capture = false,
  once = false,
  passive,
  signal,
} = typeof options === 'boolean' ? { capture: options } : (options ?? {});

const memoizedOptions = useMemo(
  () => options,
  // eslint-disable-next-line react-hooks/exhaustive-deps
  [capture, once, passive, signal],
);
```

You can find the full source code [here](https://github.com/aweebit/react-essentials/blob/e43eb9ddd60eeb3af62d8310b047d20c25d86bc2/src/hooks/useEventListener.ts#L158-L169).

This solution may not be the prettiest, but it has the additional advantage of `undefined` and `false` being treated as equal values for `capture` and `once`, which makes perfect sense since `false` is those options' default value, and so there isn't any difference in behavior whether it's provided explicitly or not.

Similarly, no differentiation is done between `false` / `true` and `{ capture: false }` / `{ capture: true }` because those are also just two different ways to encode the same settings.

An equivalent solution for `foo` with `a`, `b` and `c` properties would be this:

```ts
const { a, b, c } = foo;
// eslint-disable-next-line react-hooks/exhaustive-deps
const memoizedFoo = useMemo(() => foo, [a, b, c]);
```

And then you simply use `useStateWithDeps` like this:

```ts
const [derived, setDerived] = useState(() => deriveState(memoizedFoo), [memoizedFoo]);
```

If your `foo` object has too many (possible deeply nested) properties for this to be sustainable though, I think there is no other good solution that doesn't involve storing `foo`'s previous / "memoized" value in some state variable. But also `useImmediateEffect` wouldn't help you here, right?

However, even with this additional state, there is a much cleaner solution to the derived state problem of yours:

```ts
const [memoizedFoo, setMemoizedFoo] = useState(foo);
if (!isDeepEqual(foo, memoizedFoo)) {
  setMemoizedFoo(foo);
}

const [derived, setDerived] = useState(() => deriveState(memoizedFoo), [memoizedFoo]);
```

Just I personally would still prefer `useMemo` even in more complicated cases because unlike this state solution, it doesn't cause unnecessary renders by having the first render's result thrown away.

**Comment by jedwards1211** [comment]
No `useImmediateEffect` wouldn't help me with this.

I don't personally worry about getting confused by a `useStateWithDeps` API, I'm just wondering what the React team would think about all devs.

It would be nice to have some kind of `useMemo` that receives both the previous and current values in its computation, to avoid the double rendering issue in cases where all we can do is a deep equal comparison.

Overall I wish it were less of a hassle to do arbitrary comparisons between previous and current values of props and state in functional stateless components.

**Comment by aweebit** [comment]
> It would be nice to have some kind of `useMemo` that receives both the previous and current values in its computation, to avoid the double rendering issue in cases where all we can do is a deep equal comparison.
> 
> Overall I wish it were less of a hassle to do arbitrary comparisons between previous and current values of props and state in functional stateless components.

I was able to find some issues suggesting solutions for this:

- https://github.com/facebook/react/issues/14476
- https://github.com/facebook/react/issues/15154
- https://github.com/facebook/react/issues/28465

Probably the problem with those solutions is that again, they'd give us devs too much flexibility to the point where we could misuse them without realizing we're doing something wrong. I'd still prefer having that flexibility to using state and having those confusing and unnecessary throwaway renders as a result though, just not sure it's really that necessary after all because we can always just do this thing I suggested:

> ```ts
> const { a, b, c } = foo;
> // eslint-disable-next-line react-hooks/exhaustive-deps
> const memoizedFoo = useMemo(() => foo, [a, b, c]);
> ```
